src/Examples/MNIST.cs

```diff
@@ -70,120 +65,107 @@ static void Main(string[] args)
-                }
+            var normImage = TorchVision.Transforms.Normalize(new double[] { 0.1307 }, new double[] { 0.3081 }, device: device);
+
+            using (MNISTReader train = new MNISTReader(targetDir, "train", _trainBatchSize, device: device, shuffle: true, transform: normImage),
+                                test = new MNISTReader(targetDir, "t10k", _testBatchSize, device: device, transform: normImage)) {
 
-                sw.Stop();
-                Console.WriteLine($"Elapsed time: {sw.Elapsed.TotalSeconds} s.");
-                Console.ReadLine();
+                TrainingLoop(dataset, device, model, train, test);
             }
         }
 
-        private static Sequential GetModel(Device device)
+        internal static void TrainingLoop(string dataset, Device device, Model model, MNISTReader train, MNISTReader test)
         {
-            var seq = Sequential(
-                ("conv1", Conv2d(1, 10, 5)),
-                ("pool1", MaxPool2d(kernelSize: new long[] { 2, 2 })),
-                ("relu1", ReLU()),
-                ("conv2", Conv2d(10, 20, 5)),
-                ("dropout1", FeatureAlphaDropout()),
-                ("pool2", MaxPool2d(kernelSize: new long[] { 2, 2 })),
-                ("relu2", ReLU()),
-                ("flatten", Flatten()),
-                ("fc1", Linear(320, 64)),
-                ("relu3", ReLU()),
-                ("dropout2", Dropout()),
-                ("fc2", Linear(64, 10)),
-                ("logsm", LogSoftmax(1)));
-
-            return (device != null && device.Type == DeviceType.CUDA) ?
-                seq.to(device) as Sequential :
-                seq;
+            if (device.Type == DeviceType.CUDA) {
+                _epochs *= 4;
+            }
+
+            var optimizer = NN.Optimizer.Adam(model.parameters());
+
+            var scheduler = NN.Optimizer.StepLR(optimizer, 1, 0.7, last_epoch: 5);
+
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+
+            for (var epoch = 1; epoch <= _epochs; epoch++) {
+
+                Train(model, optimizer, nll_loss(reduction: Reduction.Mean), device, train, epoch, train.BatchSize, train.Size);
+                Test(model, nll_loss(reduction: NN.Reduction.Sum), device, test, test.Size);
+
+                Console.WriteLine($"End-of-epoch memory use: {GC.GetTotalMemory(false)}");
+            }
+
+            sw.Stop();
+            Console.WriteLine($"Elapsed time: {sw.Elapsed.TotalSeconds:F1} s.");
+
+            Console.WriteLine("Saving model to '{0}'", dataset + ".model.bin");
+            model.save(dataset + ".model.bin");
         }
 
-        private class Model : CustomModule
+        internal class Model : CustomModule
         {
-            private Conv2d conv1 = Conv2d(1, 10, 5);
-            private Conv2d conv2 = Conv2d(10, 20, 5);
-            private Linear fc1 = Linear(320, 50);
-            private Linear fc2 = Linear(50, 10);
+            private Conv2d conv1 = Conv2d(1, 32, 3);
```

Comment:
I think examples could use named parameters for things like this, it reads much more clearly and teaches